### PR TITLE
Multiple improvements / bug fixes

### PR DESCRIPTION
**Pull request by dholzmueller**
Summary

  This PR bundles fixes for five open issues, pre-training improvements, and documentation updates for the 2.2.0 release.

  Bug fixes:
  - #118 — Finetuning now supports string/categorical DataFrame columns (adds TransformToNumerical to the finetuning preprocessing path)
  - #128 — Fix PyTorch autograd error with partial module freezing during fine-tuning (use detach() at frozen→trainable module boundaries)
  - #136 — Fix DatetimeEncoder sin/cos off-by-one (divide by p, not p-1)
  - #140 — Fix float16 input arrays crashing Yeo-Johnson normalization (upcast to float32 in PreprocessingPipeline)
  - #143 — Fix predict_proba crash on all-NaN categorical columns (remove broken batch-global feature-mask detection that was intended for SHAP but didn't work with SHAP's coalition batching)

  Pre-training improvements:
  - Add bfloat16 support to --dtype (GradScaler only enabled for float16)
  - Automatically disable cuDNN SDPA backend (slower than Flash Attention on Hopper)
  - Switch v2 curriculum scripts to --dtype float16

  Documentation:
  - Remove "state-of-the-art" claims from README and docs (TabICLv2 is now described as "an open tabular foundation model")
  - Full 2.2.0 changelog with attribution for all merged PRs since 2.1.1

### CODE REVIEW

**Comment by dholzmueller** [comment]
This PR is finally ready for review, after going through many rounds of ChatGPT review. @ogrisel or @GaelVaroquaux, do you want to take a look? I'd like to release v2.2.0 soon after merging.

File: tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import torch
+import pytest
+
+# Make the tests/ directory importable so test modules can do `from conftest import ...`
+sys.path.insert(0, os.path.dirname(__file__))
+
+# Disable oneDNN/MKLDNN when requested (used on Windows CI to avoid flaky
+# "could not create a primitive" errors from the oneDNN backend).
+if os.environ.get("TABICL_DISABLE_MKLDNN") == "1":
+    torch.backends.mkldnn.enabled = False
```
Comment (ogrisel) [inline tests/conftest.py:15]: This hides a bug on the CI but we should probably at least open an issue for it on the tabicl tracker and ideally craft a minimal reproducer to report upstream (or alternatively link to an existing upstream issue if this has already been reported).

File: CHANGES.md [hunk 2]
```diff
@@ -11,20 +11,74 @@ New features
...
+
+- Improve non-CUDA GPU inference reliability and performance (including XPU): inference now consistently runs on the configured backend device, uses backend-appropriate autocast, and queries available memory plus async stream/event primitives through backend-agnostic `torch.<backend>` APIs (with safe synchronous fallbacks when async is unavailable). This fixes pathological auto-batch sizing (e.g. batch size forced to 1) and restores expected accelerated inference behavior on supported non-CUDA GPU backends. When `device=None`, estimators now default to CUDA when available, otherwise XPU, then MPS, and then CPU. ([PR#144](https://github.com/soda-inria/tabicl/pull/144))
+
+- Improve Apple Silicon MPS inference: MPS now uses the same AMP, auto-batching, and memory-aware inference path as other accelerators instead of falling back to the CPU path. `use_amp="auto"` is device-aware (off on CPU; size heuristic on CUDA/XPU/MPS), and float16 KV caches are kept on MPS when AMP is enabled. MPS is included in the default device order (CUDA → XPU → MPS → CPU). ([PR#144](https://github.com/soda-inria/tabicl/pull/144))
```
Comment (ogrisel) [inline CHANGES.md]: This part of the changelog is redundant with the previous entry.

```suggestion
- Improve Apple Silicon MPS inference: MPS now uses the same AMP, auto-batching, and memory-aware inference path as other accelerators instead of falling back to the CPU path. `use_amp="auto"` is device-aware (off on CPU; size heuristic on CUDA/XPU/MPS), and float16 KV caches are kept on MPS when AMP is enabled. ([PR#144](https://github.com/soda-inria/tabicl/pull/144))
```

File: scripts/train_v2_clf_stage1.sh [hunk 3]
```diff
@@ -29,7 +29,7 @@ torchrun --standalone --nproc_per_node=$NUM_GPUS -m tabicl.train \
             --wandb_project TabICLv2 \
             --wandb_name tabiclv2_clf_stage1 \
             --device cuda \
-            --dtype float32 \
+            --dtype float16 \
```
Comment (ogrisel) [inline scripts/train_v2_clf_stage1.sh]: Out of curiosity, what is the expected impact of this change on the tradeoff between pretraing cost / pretraining stability and predictive quality of the resulting model?

Did you re-run a full pretraining session with AMP enabled?

**Review by ogrisel — APPROVED** [review]
I did a partial review on the parts I felt confident with and only found the following nits/questions:

File: tests/conftest.py (hunk 1 above)
Comment (ogrisel) [reply tests/conftest.py:15]: Then this inline comment should point to the issue in the tabicl tracker.

File: scripts/train_v2_clf_stage1.sh (hunk 3 above)
Comment (dholzmueller) [reply scripts/train_v2_clf_stage1.sh]: The original pre-training used float16 according to Jingang. Stability should be okay thanks to GradScaler. In terms of speed, I observed up to 30% speedup of float16 compared to float32.

File: tests/conftest.py (hunk 1 above)
Comment (dholzmueller) [reply tests/conftest.py:15]: One problem is that this error is non-deterministic, e.g. it typically occured on one of the five python versions in the GitHub Windows CI but not always the same one. So the problem might be specific to some configuration details on some nodes of the GitHub CI. There are some other reports about the same error message but they might not have the same problem, e.g. https://github.com/intel/compute-runtime/issues/963.

File: CHANGES.md (hunk 2 above)
Comment (dholzmueller) [reply CHANGES.md]: Thanks, Claude merged the two points.

**Comment by dholzmueller** [comment]
@ogrisel if you have the memory optimization (chunking?) coming soon, we could also add that to v2.2.0.

**Comment by ogrisel** [comment]
> @ogrisel if you have the memory optimization (chunking?) coming soon, we could also add that to v2.2.0.

I am a bit busy to get it this week and maybe review will take a bit of time so I would not hold v2.2.0.

We can always release v2.3.0 with the memory optimizations in a few weeks.

File: scripts/train_v2_clf_stage1.sh (hunk 3 above)
Comment (Zhengli-Wang) [reply scripts/train_v2_clf_stage1.sh]: Thanks for the clarification. I agree that FP16 can reduce the training cost, and I do not dispute that the original pre-training may have used FP16.

However, in my reproduction, stability seems to depend not only on the dtype, but also critically on the qassmax path and the exact AMP configuration. My current results are:

qassmax + FP32: training is stable, but slower and more memory-intensive;
no qassmax + FP16: Stage 1 trains stably, with a stable cross-entropy loss;
qassmax + FP16: training produces NaNs;
BF16: training collapses regardless of whether qassmax is enabled.
Therefore, GradScaler alone does not seem to prevent the failure mode I observe. It may help with gradient underflow/overflow, but it does not necessarily prevent numerical issues occurring in the forward pass or inside the attention/qassmax computation.

Based on these results, I would be cautious about changing the default dtype to FP16 before running a full pre-training and downstream evaluation under the exact intended configuration. Could you share the exact commit/configuration used for the original FP16 run, including whether qassmax was enabled and whether its computation was performed under autocast or FP32?

Comment (dholzmueller) [reply scripts/train_v2_clf_stage1.sh]: Thanks for your feedback! Unfortunately we lost access to the original pre-training config, but qassmax was definitely enabled as it is contained in the published checkpoints. I didn't find any autocast modifier in the code. However, older configs (including those for the ablations) used FP32, so the FP16 is only based on what Jingang told me.

Comment (ogrisel) [reply scripts/train_v2_clf_stage1.sh]: Maybe we should switch back the scripts to float32 for the sake of stability and explore float16 pre-training later in follow-up PRs?

Comment (dholzmueller) [reply scripts/train_v2_clf_stage1.sh]: I reverted to float32 and added a note about float16 for now.
